node/src/components/consensus/era_supervisor.rs

```diff
@@ -397,32 +400,72 @@ where
         );
         let mut effects = responder.respond(signature).ignore();
         if block_header.switch_block() {
-            // TODO: Learn the new weights from contract (validator rotation).
-            let validator_stakes = self.era_supervisor.validator_stakes.clone();
-            self.era_supervisor
-                .current_era_mut()
-                .consensus
-                .deactivate_validator();
-            let new_era_id = block_header.era_id().successor();
-            info!(?new_era_id, "Era created");
-            let results = self.era_supervisor.new_era(
-                new_era_id,
-                Timestamp::now(), // TODO: This should be passed in.
-                validator_stakes,
-                block_header.timestamp(),
-                block_header.height() + 1,
-                *block_header.global_state_hash(),
+            // if the block is a switch block, we have to get the validators for the new era and
+            // create it, before we can say we handled the block
+            let request = GetEraValidatorsRequest::new(
+                (*block_header.global_state_hash()).into(),
+                block_header.era_id().successor().0,
+                ProtocolVersion::V1_0_0,
+            );
+            effects.extend(self.effect_builder.get_validators(request).event(|result| {
+                Event::GetValidatorsResponse {
+                    switch_block_header: block_header,
+                    get_validators_result: result,
+                }
+            }));
+        } else {
+            // if it's not a switch block, we can already declare it handled
+            effects.extend(
+                self.effect_builder
+                    .announce_block_handled(block_header.height())
+                    .ignore(),
             );
-            effects.extend(self.handle_consensus_results(new_era_id, results));
         }
-        effects.extend(
-            self.effect_builder
-                .announce_block_handled(block_header.height())
-                .ignore(),
-        );
         effects
     }
 
+    pub(super) fn handle_validators_response(
+        &mut self,
+        block_header: BlockHeader,
+        validators: GetEraValidatorsResult,
+    ) -> Effects<Event<I>> {
+        match validators {
+            GetEraValidatorsResult::Success { validator_weights } => {
+                let validator_stakes = validator_weights
+                    .into_iter()
+                    .map(|(key, stake)| {
+                        (
+                            key.try_into().ok().expect("keys should convert correctly"),
```

Comment:
Is this a panic a user could trigger by bonding to an invalid public key?
In that case, we should slash them, not panic. (Not in this PR, of course, but something to keep in mind.)

Reply:
I'm not sure... Don't you need to sign the deploy executing the bonding with the key you are using to bond? If yes, then the key has to be valid, or the signature won't validate correctly. But I don't know if this is the case.

Reply:
No, the deploy is signed with your account key, not with the validator public key.